core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,383 @@
+     * as a state progression as it goes to Removing.
+     *
+     * @param container the container to destroy
+     */
+    def destroyContainer(container: Container) = {
+        context.parent ! ContainerRemoved
+
+        val unpause = stateName match {
+            case Paused => container.resume()(TransactionId.invokerNanny)
+            case _      => Future.successful(())
+        }
+
+        unpause
+            .flatMap(_ => container.destroy()(TransactionId.invokerNanny))
+            .map(_ => ContainerRemoved).pipeTo(self)
+
+        goto(Removing)
+    }
+
+    /**
+     * Runs the job, initialize first if necessary.
+     *
+     * @param container the container to run the job on
+     * @param job the job to run
+     * @return a future completing after logs have been collected and
+     *         added to the WhiskActivation
+     */
+    def initializeAndRun(container: Container, job: Run)(implicit tid: TransactionId): Future[WhiskActivation] = {
+        val actionTimeout = job.action.limits.timeout.duration
+
+        // Only initialize iff we haven't yet warmed the container
+        val initialize = stateData match {
+            case data: WarmedData => Future.successful(Interval.zero)
+            case _                => container.initialize(job.action.containerInitializer, actionTimeout)
+        }
+
+        val activation: Future[WhiskActivation] = initialize.flatMap { initInterval =>
+            val parameters = job.msg.content getOrElse JsObject()
+
+            val environment = JsObject(
+                "api_key" -> job.msg.user.authkey.compact.toJson,
+                "namespace" -> job.msg.user.namespace.toJson,
+                "action_name" -> job.msg.action.qualifiedNameWithLeadingSlash.toJson,
+                "activation_id" -> job.msg.activationId.toString.toJson,
+                // compute deadline on invoker side avoids discrepancies inside container
+                // but potentially under-estimates actual deadline
+                "deadline" -> (Instant.now.toEpochMilli + actionTimeout.toMillis).toString.toJson)
+
+            container.run(parameters, environment, actionTimeout)(job.msg.transid).map {
+                case (runInterval, response) =>
+                    val initRunInterval = Interval(runInterval.start.minusMillis(initInterval.duration.toMillis), runInterval.end)
+                    ContainerProxy.constructWhiskActivation(job, initRunInterval, response)
+            }
+        }.recover {
+            case InitializationError(interval, response) =>
+                ContainerProxy.constructWhiskActivation(job, interval, response)
+        }
+
+        // Sending active ack and storing the activation are concurrent side-effects
+        // and do not block further execution of the future. They are completely
+        // asynchronous.
+        activation.andThen {
+            case Success(activation) => sendActiveAck(tid, activation)
```

Comment:
is the recover broad enough to ensure active ack response on all failures?
(This is related to your past complaints that init/run complete the future with success invariably - or at least that's the idea.)